src/main/java/com/rackspacecloud/metrics/rollup/domain/Metric.java

```diff
@@ -1,118 +1,151 @@
+    public Map<String, RollupBucket<Double>> rolledUpFValues; // Keeps rolled up fvalue
+
+    @Data
+    public static class RollupBucket<T extends Number> {
+        private T min;
+        private T max;
+        private T mean;
+
+        private T sum;
+        private long count;
+    }
+
     /**
      * Define the rollup-key that is used to aggregate the data.
      * @return rollup-key
      */
     @Override
-    public String getRollupKey(){
-        StringBuilder sb = new StringBuilder();
-        sb.append(String.format("%s", getSystemMetadata()));
-        this.rollupKey = sb.toString();
+    public String getRollupKey() {
+        String[] rollupKeyTags = new String[] {
+                accountType, account, monitoringSystem, collectionName,
+                device, deviceLabel, collectionLabel
+        };
 
-        // TODO: Check what else needs to be part of the rollup key
+        this.rollupKey = String.join(".", rollupKeyTags);
 
-        return rollupKey;
-    }
+        // TODO: Check what else needs to be part of the rollup key
 
-    private String getSystemMetadata(){
-        return String.format("%s.%s.%s.%s.%s",
-                systemMetadata.get("tenantId"),
-                systemMetadata.get("checkType"),
-                systemMetadata.get("accountId"),
-                systemMetadata.get("entityId"),
-                systemMetadata.get("checkId")
-        );
+        return this.rollupKey;
     }
 
     @Override
     public Metric reduce(Metric newValue) {
-        reduceIValues(newValue);
-        reduceFValues(newValue);
+        reduceIValues(newValue.getIvalues());
+        reduceFValues(newValue.getFvalues());
         return this;
     }
 
-    private void reduceIValues(Metric newValue) {
-        if(iValuesForRollup == null) {
-            iValuesForRollup = new HashMap<>();
-            ivalues.forEach((k,v) -> {
-                List<Long> valueList = new ArrayList<>();
-                valueList.add(v);
-                iValuesForRollup.put(k, valueList);
-            });
+    private void reduceIValues(Map<String, Long> newIValues) {
+        if(rolledUpIValues == null) {
+            rolledUpIValues = new HashMap<>();
         }
 
-        Map<String, Long> newIValues = newValue.ivalues;
-
-        for(String iKey : newIValues.keySet()){
-            if(this.iValuesForRollup.containsKey(iKey)){
-                List<Long> iValuesList = this.iValuesForRollup.get(iKey);
-                iValuesList.add(newIValues.get(iKey));
+        newIValues.forEach((k, v) -> {
+            if(!rolledUpIValues.containsKey(k)) { // Initialize rollup bucket for given key
+                RollupBucket<Long> rollupBucket = new RollupBucket<>();
+                rollupBucket.min = v;
+                rollupBucket.max = v;
+                rollupBucket.sum = v;
+                rollupBucket.count = 1;
+                rollupBucket.mean = rollupBucket.sum/rollupBucket.count;
```

Comment:
Isn't this also `v` for a new bucket?

Reply:
I am assuming, your "this" is for `rollupBucket.sum` or `rollupBucket.mean`. If that's the case, then you are right, but it's `v` anyway based on the operation.

Reply:
I meant this could be `rollupBucket.mean = v;` which is simpler; the way it's right now it's more self-documenting. Probably doesn't matter much either way.